compiler/rustc_trait_selection/src/traits/select/mod.rs

```diff
@@ -214,6 +213,122 @@ enum BuiltinImplConditions<'tcx> {
     Ambiguous,
 }
 
+fn tys_may_be_equivalent<'tcx>(oblig_ty: Ty<'tcx>, impl_ty: Ty<'tcx>) -> bool {
+    let k = impl_ty.kind();
+
+    // First, look for obvious equivalences, e.g. the types are equal. The most
+    // common case is to find such an equivalence trivially, e.g. `ty::Bool`
+    // and `ty::Bool`.
+    let mut maybe_equiv = match oblig_ty.kind() {
```

Comment:
I prefer one large `match (oblig_ty.kind(), impl_ty.kind())` here, starting with
```rust
(_, ty::Param(_) | ty::Projection(_) | ty::Error(_)) => true,
```

Reply:
This is what I was referring to when I said I chose to structure it with exhaustive matches. Using `_` here makes it easy to overlook updating this when new variants are added to `Ty`.